### PR TITLE
Add columns for process autogroup identifier and nice value

### PR DESCRIPTION
**Pull request by natoscott**
Adds AUTO (autogroup) and ANI (autogroup nice) columns that
report the information from /proc/PID/autogroup, as well as
handlers for '{' and '}' to change the autogroup nice value.

This is guarded by /proc/sys/kernel/sched_autogroup_enabled
such that sampling and/or changing values wont be attempted
unless the kernel feature is enabled.

Fixes: #720

### CODE REVIEW

**Review by fasterit — APPROVED** [review]

**Comment by BenBE** [comment]
What about `AGRP` instead of `AUTO`? IMHO more descriptive random sequence of letters ;-)

File: linux/LinuxProcess.c [hunk 1]
```diff
@@ -159,6 +161,40 @@ bool LinuxProcess_setIOPriority(Process* this, Arg ioprio) {
    return (LinuxProcess_updateIOPriority((LinuxProcess*)this) == ioprio.i);
 }
 
+bool LinuxProcess_isAutogroupEnabled(void) {
+   FILE* file = fopen(PROCDIR "/sys/kernel/sched_autogroup_enabled", "r");
+   if (!file)
+      return false;
+   unsigned int enabled;
+   int count = fscanf(file, "%u", &enabled);
+   fclose(file);
```
Comment (BenBE) [inline linux/LinuxProcess.c:170]:
```suggestion
   char buf[16];
   if (xReadfile(PROCDIR "/sys/kernel/sched_autogroup_enabled", buf, sizeof(buf)) < 0)
      return false;

   unsigned int enabled;
   int count = sscanf(buf, "%u", &enabled);
```

File: linux/LinuxProcess.h [hunk 2]
```diff
@@ -99,6 +100,10 @@ typedef struct LinuxProcess_ {
    unsigned long ctxt_diff;
    char* secattr;
    unsigned long long int last_mlrs_calctime;
+
+   /* Autogroup scheduling (CFS) information */
+   unsigned int autogroup;
```
Comment (BenBE) [inline linux/LinuxProcess.h:105]: `autogroup_id` / `autogroup_gid`?

File: linux/LinuxProcessList.c [hunk 3]
```diff
@@ -909,6 +909,24 @@ static void LinuxProcessList_readOomData(LinuxProcess* process, openat_arg_t pro
    fclose(file);
 }
 
+static void LinuxProcessList_readAutogroup(LinuxProcess* process, openat_arg_t procFd) {
+   FILE* file = fopenat(procFd, "autogroup", "r");
```
Comment (BenBE) [inline linux/LinuxProcessList.c:913]: Use `xReadfileAt`?

File: linux/Platform.c [hunk 4]
```diff
@@ -150,8 +150,41 @@ static Htop_Reaction Platform_actionSetIOPriority(State* st) {
    return HTOP_REFRESH | HTOP_REDRAW_BAR | HTOP_UPDATE_PANELHDR;
 }
 
+
```
Comment (BenBE) [inline linux/Platform.c:153]: No need for double blank here …

File: linux/ProcessField.h [hunk 5]
```diff
@@ -44,8 +44,10 @@ in the source distribution for its full text.
    M_PSSWP = 121,                \
    CTXT = 122,                   \
    SECATTR = 123,                \
+   AUTOGROUP = 127,              \
+   AUTOGROUP_NICE = 128,         \
                                  \
-   DUMMY_BUMP_FIELD = CWD,       \
+   DUMMY_BUMP_FIELD = 128,       \
```
Comment (BenBE) [inline linux/ProcessField.h:50]: Can be dropped, was only there to ensure the last ID defined here was the overall highest ID available.

File: linux/LinuxProcess.c [hunk 6]
```diff
@@ -100,6 +100,8 @@ const ProcessFieldData Process_fields[LAST_PROCESSFIELD] = {
    [PROC_COMM] = { .name = "COMM", .title = "COMM            ", .description = "comm string of the process from /proc/[pid]/comm", .flags = 0, },
    [PROC_EXE] = { .name = "EXE", .title = "EXE             ", .description = "Basename of exe of the process from /proc/[pid]/exe", .flags = 0, },
    [CWD] = { .name = "CWD", .title = "CWD                       ", .description = "The current working directory of the process", .flags = PROCESS_FLAG_CWD, },
+   [AUTOGROUP] = { .name = "AUTOGROUP", .title = "AUTO", .description = "The autogroup identifier of the process", .flags = PROCESS_FLAG_LINUX_AUTOGROUP, },
```
Comment (BenBE) [inline linux/LinuxProcess.c:103]: `AUTOGROUP` -> `AUTOGROUP_ID` as there's multiple fields on this topic.

**Review by BenBE — CHANGES_REQUESTED** [review]

File: linux/LinuxProcess.c [hunk 7]
```diff
@@ -159,6 +161,40 @@ bool LinuxProcess_setIOPriority(Process* this, Arg ioprio) {
...
+   xSnprintf(buffer, sizeof(buffer), PROCDIR "/%d/autogroup", this->pid);
+
+   FILE* file = fopen(buffer, "r+");
+   if (!file)
+      return false;
+
+   unsigned int identity;
+   long int nice;
+   bool success;
+   int ok = fscanf(file, "/autogroup-%u nice %ld", &identity, &nice);
```
Comment (BenBE) [inline linux/LinuxProcess.c:185]: Type mismatch with kernel source:
https://github.com/torvalds/linux/blob/v5.13/kernel/sched/autogroup.c#L254

**Review by BenBE — CHANGES_REQUESTED** [review]

File: linux/LinuxProcess.c (hunk 1 above)
Comment (cgzones) [reply linux/LinuxProcess.c:170]: Then we would not even need the `scanf()`, as we could just check `buf[0]` (which should be either '0' or '1').

File: linux/LinuxProcessList.c [hunk 8]
```diff
@@ -909,6 +909,24 @@ static void LinuxProcessList_readOomData(LinuxProcess* process, openat_arg_t pro
    fclose(file);
 }
 
+static void LinuxProcessList_readAutogroup(LinuxProcess* process, openat_arg_t procFd) {
```
Comment (cgzones) [inline linux/LinuxProcessList.c:912]: Maybe set `process->autogroup` and `process->autogroup_nice` to some invalid value first and print `N/A` in `LinuxProcess_writeField()` on those?

File: linux/LinuxProcessList.c [hunk 9]
```diff
@@ -1323,6 +1341,15 @@ static bool LinuxProcessList_recurseProcTree(LinuxProcessList* this, openat_arg_
       return false;
    }
 
+   bool autogroup = (settings->flags & PROCESS_FLAG_LINUX_AUTOGROUP);
+   if (autogroup) {
+      // Refer to sched(7) 'autogroup feature' section
+      // The kernel feature can be enabled/disabled through procfs at
+      // any time, so check for it at the start of each sample - only
+      // read from per-process procfs files if it's globally enabled.
+      autogroup = LinuxProcess_isAutogroupEnabled();
```
Comment (cgzones) [inline linux/LinuxProcessList.c:1350]: This leads to `LinuxProcess_isAutogroupEnabled()` being called for each process and thread if `PROCESS_FLAG_LINUX_AUTOGROUP` is set.

**Comment by natoscott** [comment]
> What about `AGRP` or `AGID` instead of `AUTO`? IMHO more descriptive random sequence of letters ;-)

Done, went with AGRP - thanks.

**Review by BenBE — APPROVED** [review]

**Pull request merged**